### ISSUE TITLE
[Feature] Run Playwright tests from Typescript code using CLI function.

### ISSUE THREAD

**Issue by deepakkamboj** [microsoft/playwright#8971]
We are currently using Jest for running playwright test cases. So this code is working perfectly fine for us:
 
```ts
import { runCLI } from "jest";
 const { results } = await runCLI(
      {
        $0: "",
        config: jestConfigPath,
        ...options,
        _: testNamePatterns,
      },
      [integrationTestsPackagePath]
    );

```
 
We are getting test results and we can decide the next steps for test run. 
We would like to switch to playwright-test-runner.
We would like to create playwright.config.ts on runtime and then run playwright tests cases using similar command above for playwright test runner.

What we want is to somehow expose this line of code (#154):
 
const result = await runner.run(!!opts.list, filePatternFilters, opts.project || undefined);

https://github.com/microsoft/playwright/blob/eafba43e15bcca8d10da30bc7c53d4092bb2374a/src/test/cli.ts#L154
 
![image](https://user-images.githubusercontent.com/12835833/133660151-7372ed16-cb8f-4f12-9e78-4be0720fadac.png)

This function should be exposed so that we can consume it in our Typescript code to run playwright test cases using playwright test runner.

**Comment by mxschmitt** [comment]
Seems like a duplicate of #7275.

**Comment by deepakkamboj** [comment]
We have repo which has around 300 packages and 10 apps.

We have developed a separate playwright-solution package to run integration tests on these apps and packages.

We have custom code to identify the code changes done in various apps and packages. Based on affected apps and packages, we identify the various test directories containing integration tests. We then run integration tests using our custom function runPlaywrightTests() against each of these test directories.

const results = await runPlaywrightTests (integrationTestPackagePath();

runPlaywrightTests creates a jest.config.ts file at the runtime based on integrationTestPackagePath, and other options (like broser, headless, runType, envTyes, etc.)

then we call the Jest CLI command to run test cases which in turn return the test results:

const { results } = await runCLI(
{
$0: "",
config: jestConfigPath,
...options,
_: testNamePatterns,
},
[integrationTestsPackagePath]
);

We have similar requirement in multiple repositories where we have to execute Playwright test cases programmatically.

**Comment by deepakkamboj** [comment]
We have to run various tests based on type of build pipeline executing them. 

e.g. 

- PrTests runs on PR and CI Build Pipeline. These test uses mockes.
- Nightly Mock Tests - runs on nightly build pipeline uses Mocks again.
- Nightly Live Tests - runs on nightly build pipeline uses LIVE servers and services.
- UI Runner Tests - runs on scheduled using LIVE servers and services.

Then we have to also decide whether a particular run on which type of environments:

- Dev
- Test
- PreProd
- Preview
- Prod

**Comment by pavelfeldman** [comment]
A couple of things you can try while we are considering the API for this:

- Run @playwright/test in a subprocess (child_process.spawn), pass the generated config as `-c` into it
- Use your own [reporter](https://playwright.dev/docs/api/class-reporter) that would get detailed information on the test status

All in all should be few lines of code that address your use case.

**Comment by deepakkamboj** [comment]
I have already using this workaround:

async function systemSync(cmd: string): Promise<Number> {
  let exitCode = 0;
  await child_process
    .exec(cmd, (err, stdout, stderr) => {
      console.log(stdout);
      console.log(stderr);

      if (err && err.code) {
        console.log(err);
        console.log('Exit Code: ', err.code);
        console.error(`##[error]Playwright Integration Tests - FAILED`);
      } else {
        console.log(`##[section]Playwright Integration Tests - COMPLETED`);
      }
    })
    .on('exit', (code) => {
      console.log('Final exit code is: ', code);
      if (code) {
        exitCode = code;
      }
    });

  return exitCode;
}
async function run() {
  const rootDir = path.join(process.cwd(), 'tests');

  console.log('Current Working Directory: ', rootDir);

  try {
    const parameters = await getCommandParameters();
    const browser = parameters.browser;
    const headless = parameters.headless;
    const group = parameters.group;

    console.log('Input CLI Parameters: ', parameters);

    const exitCode = await systemSync('npx playwright test');
  } catch (e) {
    console.error('Error: Exception occurred while running Playwright test cases: ', e.message);
  }
}

But this is not providing all the benefits which runCLI from Jest provides.

**Comment by pavelfeldman** [comment]
> But this is not providing all the benefits which runCLI from Jest provides.

Being like Jest is not a priority to us, but satisfying out customers is. It is not clear from your request what exact benefits you are referring to.

**Comment by softleo-llc** [comment]
We are having similar requirement where we want to run the Playwright command programmatically. We have custom CLI arguments. Based on these arguments, we would like to create Playwright config on runtime and then run playwright test cases.

We have to also configure various configurations like @grep for filtering the test cases and using mock or live services, etc.

**Comment by deepakkamboj** [comment]
@mxschmitt - Please let us know when this feature will be available. This is the requirement from lot of teams but it's getting pushed from v1.16 to v1.18.

**Comment by ArturT** [comment]
Our users at [Knapsack Pro](https://knapsackpro.com/?utm_source=github&utm_medium=github-issue&utm_campaign=playwright-runcli-issue) are asking for Playwright support and we would like to integrate our tool with Playwright so having a similar feature as Jest runCLI would be very helpful.

We would like to let users leverage our [Queue Mode API to distribute tests in parallel in a dynamic way across parallel machines](https://docs.knapsackpro.com/2020/how-to-speed-up-ruby-and-javascript-tests-with-ci-parallelisation#dynamic-tests-split---queue-mode) to run fast CI builds. 

For Jest, we use [`jest.runCLI`](https://jestjs.io/docs/cli) (how [jest.runCLI is integrated in `@knapsack-pro/jest`](https://github.com/KnapsackPro/knapsack-pro-jest/blob/1b43784d87f874ee270030fa9063e752ba61695e/src/knapsack-pro-jest.ts#L46)) to run a given set of test files and collect information about how long does it take to run test files. We collect information about the execution time of each test file in Knapsack Pro API.

For Cypress we use [`cypress.run`](https://docs.cypress.io/guides/guides/command-line) ([how `cypress.run` is integrated with `@knapsack-pro/cypress`](https://github.com/KnapsackPro/knapsack-pro-cypress/blob/1aa0eb710b524a94af53ab72c504736ef29548cb/src/knapsack-pro-cypress.ts#L48)).


Basically, it would be great to see Playwright having `runCLI` option:
* so that we can pass custom options (all available options supported by Playwright) and a list of test files
* we could get results about executed tests to check how long does it take to run each test file in seconds
* we could run Playwright `runCLI` multiple time within the process. We pass a set of test files to it. Get the results about execution time and then start another Playwright `runCLI` command to run 2nd set of test files fetched from Knapsack Pro Queue API and so on. Here I described [how we do it for Cypress](https://docs.knapsackpro.com/2020/how-to-build-native-integration-with-knapsack-pro-api-to-run-tests-in-parallel-for-any-test-runner-testing-framework).

Thanks

**Comment by deepakkamboj** [comment]
@pavelfeldman - Please let us know when this feature will be available. This is the requirement from a lot of teams but it's getting pushed from v1.16 to v1.19 now. We have been waiting for this feature to be available for long time now.

**Comment by pavelfeldman** [comment]
I'd recommend to not depend on the progress of this one. The API for running the tests is going to be out-of-process, you can copy it from https://github.com/microsoft/playwright-vscode/blob/8fa23f5373c8f83cbf09b3cb85295829e8229730/src/oopReporter.ts and adjust it to liking. It is unlikely we follow up with the official API for it.

**Comment by deepakkamboj** [comment]
The link you shared is for reporter. How can reporter will help in running the playwright tests programatically?

**Comment by Meir017** [comment]
> The link you shared is for reporter. How can reporter will help in running the playwright tests programatically?

the reporter is used from playwrightTest
here https://github.com/microsoft/playwright-vscode/blob/8fa23f5373c8f83cbf09b3cb85295829e8229730/src/playwrightTest.ts#L101-L106

**Comment by mxschmitt** [comment]
You need to run Playwright Test as a sub process and pass the supported CLI parameters.

**Comment by deepakkamboj** [comment]
@mxschmitt - Please share an example.

**Comment by mxschmitt** [comment]
> @mxschmitt - Please share an example.

https://github.com/microsoft/playwright-vscode/blob/8fa23f5373c8f83cbf09b3cb85295829e8229730/src/playwrightTest.ts#L102

Basically just spawn 'npx playwright test' with your custom reporter or if sufficient '--reporter=json'.
